include/frozen/unordered_map.h

```diff
@@ -146,8 +146,60 @@ class unordered_map {
   constexpr key_equal key_eq() const { return equal_; }
 
 private:
-  constexpr auto const &lookup(Key const &key) const {
-    return items_[tables_.lookup(key)];
+  /*
+   * Template used to reduce code duplication between const and non-const
+   * functions. https://stackoverflow.com/a/62215403/4832300
+   */
+  template <class This>
+  static constexpr auto& at_impl(This& self, Key const &key) {
+    auto& kv = self.lookup(key);
+    if (self.equal_(kv.first, key))
+      return kv.second;
+    else
+      FROZEN_THROW_OR_ABORT(std::out_of_range("unknown key"));
+  }
+
+  /*
+   * Template used to reduce code duplication between const and non-const
+   * functions. https://stackoverflow.com/a/62215403/4832300
+   */
+  template <class This>
+  static constexpr auto find_impl(This& self, Key const &key) {
+    auto& kv = self.lookup(key);
+    if (self.equal_(kv.first, key))
+      return &kv;
+    else
+      return self.items_.end();
+  }
+
+  /*
+   * Template used to reduce code duplication between const and non-const
+   * functions. https://stackoverflow.com/a/62215403/4832300
+   */
+  template <class This>
+  static constexpr auto equal_range_impl(This& self, Key const &key) {
+    auto& kv = self.lookup(key);
+    using kv_ptr = decltype(&kv);
+    if (self.equal_(kv.first, key))
+      return std::pair<kv_ptr, kv_ptr>{&kv, &kv + 1};
```

Comment:
I agree, I far prefer that syntax, but I cannot get it to work.

The issue is that the return value changes if `This` is const pointer or not. Doing `return {&kv, &kv + 1}` results in the compiler complaining it cannot determine the return type.

Trailing return types also do not work, since the return type is based on `decltype(&kv)` and `kv` cannot be used in that context.